Sunghcho/denormals

hanbitmyths opened the pull request:
**Description**: Add new session option and global threading option to set denormal as zero.

**Motivation and Context**
When a model contains denormal floating number in weight or input, computation cost is high and sometimes it's more than 30 times slower than without denormals. This PR introduces new session option and global threading option to be able to set denormal as zero.
One note is that flush-to-zero and Denormal-as-zero can be done at thread level. It means main thread and OpenMP threads can't be fully controlled by session option. only the first session option is effective to main thread and OpenMP threads.

tracysh reviewed (changes requested):
We shouldn't need to pull in cpuinfo just to get the DAZ flag: we can add something to MLAS that exposes a bool for you to check instead.

yuslepukhin commented on include/onnxruntime/core/session/onnxruntime_c_api.h:
```diff
@@ -1088,7 +1088,13 @@ struct OrtApi {
+   * Use this API to configure the global thread pool options to be used in the call to CreateEnvWithGlobalThreadPools.
+   * Set denormal as zero.
+   */
+  ORT_API2_STATUS(SetGlobalDenormalAsZero, _Inout_ OrtThreadingOptions* tp_options);
```
>ORT_API2_STATUS(SetGlobalDenormalAsZ [](start = 2, length = 36)

New APIs can only be appended, never inserted to preserve ABI compatibility with C# and potentially others

hanbitmyths replied:
Mistake on merging

hanbitmyths commented:
> We shouldn't need to pull in cpuinfo just to get the DAZ flag: we can add something to MLAS that exposes a bool for you to check instead. We already do CPU detection from mlas\lib\platform.cpp, we can extend that.

@tracysh, I removed cpuinfo submodule and modified to detect only sse3 support like Tensorflow.

hanbitmyths commented:
/azp run Linux CPU CI Pipeline,Linux CPU x64 NoContribops CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,MacOS CI Pipeline,MacOS NoContribops CI Pipeline,Windows CPU CI Pipeline,Linux OpenVINO CI Pipeline,Windows GPU CI Pipeline,Windows GPU TensorRT CI Pipeline,centos7_cpu

hanbitmyths commented:
/azp run Linux CPU CI Pipeline

pranavsharma commented on onnxruntime/core/common/denormal.cc:
```diff
@@ -0,0 +1,67 @@
+  explicit DenormalAsZeroInitializer(bool on) : on_(on) {}
+
+  DenormalAsZeroInitializer(const DenormalAsZeroInitializer& init) : on_(init.on_) {
+    SetDenormalAsZero(on_);
```
A bit confused. Why does the copy ctor call SetDenormalAsZero, but not the ctor?

hanbitmyths replied:
It's a trick solution to run initialization code at all open mp threads. I'll write down comments here.

pranavsharma commented on include/onnxruntime/core/session/onnxruntime_c_api.h:
```diff
@@ -1094,6 +1094,12 @@ struct OrtApi {
+
+  /**
+   * Use this API to configure the global thread pool options to be used in the call to CreateEnvWithGlobalThreadPools.
+   * Set denormal as zero.
```
Would be good to document: 1. when would someone want to call this API, what alternatives exist and what's the impact of using this (like loss of accuracy)? 2. this applies to both inter and intra op threadpools.

hanbitmyths replied:
Add details here.

pranavsharma commented on onnxruntime/core/session/inference_session.cc:
```diff
@@ -202,6 +203,20 @@ void InferenceSession::ConstructorCommon(const SessionOptions& session_options,
+#ifdef _OPENMP
+      InitializeWithDenormalAsZero(set_denormal_as_zero);
+#endif
+      SetDenormalAsZero(set_denormal_as_zero);
```
Would be good to log a line at INFO level saying that this setting is being turned on.

pranavsharma commented:
Is the flatbuffers related change intended?

hanbitmyths commented:
> Is the flatbuffers related change intended?

It's from merge, not from my change.

tracysh commented on onnxruntime/core/common/denormal.cc:
```diff
@@ -0,0 +1,67 @@
+
+void InitializeWithDenormalAsZero(bool on) {
+  DenormalAsZeroInitializer init(on);
+#pragma omp parallel for firstprivate(init)
```
From what I've seen with MSVC + openmp, if the parent thread that spawns off a parallel section does anything to change the number of threads used, then openmp recreates the task set. And the pool is per-thread, but I guess I need to see the comments describing how this works above.

hanbitmyths replied:
With firstprivate clause, all openmp threads are initialized with passing variable's copy constructor. It works even when loop count is less than maximum number of openmp threads. I put comments to explain a behavior.

tracysh commented on onnxruntime/test/perftest/command_args_parser.cc:
```diff
@@ -46,16 +46,17 @@ namespace perftest {
       "\t-u [optimized_model_path]: Specify the optimized model path for saving.\n"
       "\t-d [cudnn_conv_algorithm]: Specify CUDNN convolution algothrithms: 0(benchmark), 1(heuristic), 2(default). \n"
       "\t-q: [CUDA only] use separate stream for copy. \n"
+      "\t-z: Set denormal as zero. When tuning on this option reduces latency dramatically, a model may have denormals.\n"
```
tuning->turning?

hanbitmyths replied:
I also found the same typo from other places. Thanks for pointing out.

hanbitmyths commented on onnxruntime/core/common/denormal.cc:
```diff
@@ -0,0 +1,67 @@
+#include <intrin.h>
+#define DENORMAL_INTRINC
+// intrins headers at gcc 4.8 and older are not usable without compiler flags
+#elif defined(__GNUC__) && ((__GNUC__ >= 5) || ((__GNUC__ == 4) && (__GNUC_MINOR__ > 8)))
```
Add __clang__ not to be filtered by gcc 4.8 constraint. Clang works properly with intrinc.

tracysh commented on onnxruntime/core/common/denormal.cc:
```diff
@@ -0,0 +1,74 @@
+#endif
+#endif
+
+#include <mutex>
```
nit: is <mutex> still needed? I think you had it from std::call_once in an earlier iteration.

hanbitmyths replied:
removed

tracysh commented on include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h:
```diff
@@ -431,6 +432,7 @@ class ThreadPoolTempl : public onnxruntime::concurrency::ExtendedThreadPoolInter
 
     worker_data_.resize(num_threads_);
     for (int i = 0; i < num_threads_; i++) {
+      worker_data_[i].set_denormal_as_zero = thread_options.set_denormal_as_zero;
```
(Sorry had this comment earlier, but I don't see what happened to it): Can set_denormal_as_zero just be a flag at the same level as allow_spinning_ and not need to be in the worker data?

hanbitmyths replied:
Good point. Fixed.
